bitnami/thanos/templates/storegateway/service-sharded.yaml

```
@@ -0,0 +1,67 @@
+{{- if and .Values.storegateway.enabled .Values.storegateway.sharded.enabled }}
+
+{{- $shards := int 0 }}
+
+{{- if .Values.storegateway.sharded.hashpartitioning.shards }}
+  {{- $shards = int .Values.storegateway.sharded.hashpartitioning.shards }}
+{{- else }}
+  {{- $shards = len .Values.storegateway.sharded.timepartitioning }}
+{{- end }}
+
+{{- range $index, $_ := until $shards }}
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ include "common.names.fullname" . }}-storegateway
+  labels: {{- include "common.labels.standard" . | nindent 4 }}
+    app.kubernetes.io/component: storegateway
+    {{- if and .Values.metrics.enabled .Values.metrics.serviceMonitor.enabled }}
+    prometheus-operator/monitor: 'true'
+    {{- end }}
+    {{- if .Values.commonLabels }}
+    {{- include "common.tplvalues.render" ( dict "value" .Values.commonLabels "context" $ ) | nindent 4 }}
+    {{- end }}
+  {{- if .Values.storegateway.service.annotations }}
+  annotations: {{- include "common.tplvalues.render" (dict "value" .Values.storegateway.service.annotations "context" $) | nindent 4 }}
+  {{- end }}
```

Be careful here.. The context changes since you're in a loop:
```suggestion
  name: {{ include "common.names.fullname" $ }}-storegateway
  labels: {{- include "common.labels.standard" $ | nindent 4 }}
    app.kubernetes.io/component: storegateway
    {{- if and $.Values.metrics.enabled $.Values.metrics.serviceMonitor.enabled }}
    prometheus-operator/monitor: 'true'
    {{- end }}
    {{- if $.Values.commonLabels }}
    {{- include "common.tplvalues.render" ( dict "value" $.Values.commonLabels "context" $ ) | nindent 4 }}
    {{- end }}
  {{- if $.Values.storegateway.service.annotations }}
  annotations: {{- include "common.tplvalues.render" (dict "value" $.Values.storegateway.service.annotations "context" $) | nindent 4 }}
  {{- end }}
```
This applies to the whole file.